ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -435,6 +462,48 @@ private byte[] serializeValue(
     }
   }
 
+  /**
+   * Add missing required fields to the passed {@code FormatInfo} and return a
+   * new {@code FormatInfo} with the new fields. These fields are required to serialize the
+   * key and value schema correctly. For instance, if running INSERT on a stream with a SR schema
+   * name different from the default name used in Connect (i.e. ConnectDefault1 for Protobuf).
+   * </p>
+   * Note: I initially thought of injecting the SCHEMA_FULL_NAME property in the WITH clause
+   * when creating the stream, keep that property in the metastore and use it here instead
+   * of looking at the SR directly. But this is not compatible with existing streams because
+   * they were created previous to this fix. Those previous streams would fail with INSERT.
+   * The best option was to dynamically look at the SR schema during an INSERT statement.
+   */
+  private static FormatInfo addSerializerMissingFormatFields(
+      final SchemaRegistryClient srClient,
+      final FormatInfo formatInfo,
+      final String topicName,
+      final boolean isKey
+  ) {
+    // Only SR information, such as schema ids and full names are required, which are available
+    // for SR formats.
+    final Format format = FormatFactory.fromName(formatInfo.getFormat());
+    if (!format.supportsFeature(SerdeFeature.SCHEMA_INFERENCE)) {
+      return formatInfo;
+    }
+
+    final Set<String> supportedProperties = format.getSupportedProperties();
+
+    final ImmutableMap.Builder propertiesBuilder = ImmutableMap.builder();
+    propertiesBuilder.putAll(formatInfo.getProperties());
+
+    // The FULL_SCHEMA_NAME allows the serializer to choose the schema definition
+    if (supportedProperties.contains(ConnectProperties.FULL_SCHEMA_NAME)) {
+      if (!formatInfo.getProperties().containsKey(ConnectProperties.FULL_SCHEMA_NAME)) {
+        SchemaRegistryUtil.getParsedSchema(srClient, topicName, isKey).map(ParsedSchema::name)
+            .ifPresent(schemaName ->
+                propertiesBuilder.put(ConnectProperties.FULL_SCHEMA_NAME, schemaName));
+      }
```

Comment:
I made an improvement on the latest commit. I obtained the key/value schema full name during creation and keep it in the metastore to avoid hurting the performance. This woks only for new streams and tables, though. When users upgrade their ksql, the older streams will check for SR to get the key/value full name. It's good user experience to not make users re-create their streams. Btw, ksql uses a cached SR client, so consecutive INSERT calls won't get bad performance. 

@colinhicks any thoughts?

Reply:
I think it's fine for pre-existing objects to have reduced insert performance. We don't promote INSERT ... VALUES as an optimized write path.